### PR TITLE
Fixes #23886: Add product content to subs detail

### PR DESCRIPTION
**Pull request by akofink**
This recreates the following page in the new react subscriptions page:

screenshot from 2018-07-17 11-07-50

### CODE REVIEW

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-import { Grid, Row, Col } from 'patternfly-react';
+import { Nav, NavItem, TabPane, TabContent, TabContainer, Grid, Row, Col } from 'patternfly-react';
 import BreadcrumbsBar from 'foremanReact/components/BreadcrumbBar';
 import SubscriptionDetailInfo from './SubscriptionDetailInfo';
 import SubscriptionDetailAssociations from './SubscriptionDetailAssociations';
 import SubscriptionDetailProducts from './SubscriptionDetailProducts';
+import SubscriptionDetailEnabledProducts from './SubscriptionDetailEnabledProducts';
 import { LoadingState } from '../../../move_to_pf/LoadingState';
 import { notify } from '../../../move_to_foreman/foreman_toast_notifications';
+import { Table } from '../../../move_to_foreman/components/common/table';
```
Comment (johnpmitsch) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: Is this used or planned to be used? Not trying to nitpick, just curious :)

File: webpack/scenes/Subscriptions/Details/index.js [hunk 2]
```diff
@@ -8,6 +8,7 @@ import SubscriptionDetails from './SubscriptionDetails';
 // map state to props
 const mapStateToProps = state => ({
   subscriptionDetails: state.katello.subscriptionDetails,
+  enabledProducts: state.katello.enabledProducts,
```
Comment (johnpmitsch) [inline webpack/scenes/Subscriptions/Details/index.js]: I'm wondering if we should make` enabledProducts` part of `subscriptionDetails`? This may make more sense as the products will be related to the subscription information, though I haven't looked at an example response yet.

**Comment by johnpmitsch** [comment]
overall approach looks good! :+1: I just have an open question about the redux state structuring. Also, you will need tests if that already isn't on your todo ;)

File: webpack/scenes/Subscriptions/Details/index.js (hunk 2 above)
Comment (akofink) [reply webpack/scenes/Subscriptions/Details/index.js]: it actually is put on the subscriptionDetails. I guess this line isn't needed

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailActions.js [hunk 3]
```diff
@@ -20,4 +23,24 @@ export const loadSubscriptionDetails = subscriptionId => (dispatch, getState) =>
     .catch(result => dispatch(apiError(SUBSCRIPTION_DETAILS_FAILURE, result)));
 };
 
+export const loadEnabledProducts = subscriptionId => (dispatch, getState) => {
```
Comment (amirfefer) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailActions.js]: `getState` is redundant here

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { notify } from '../../../move_to_foreman/foreman_toast_notifications';
+
+class SubscriptionDetailEnabledProducts extends Component {
```
Comment (amirfefer) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: When no react lifecycle or the component state is involved, it better to use [stateless component](https://hackernoon.com/react-stateless-functional-components-nine-wins-you-might-have-overlooked-997b0d933dbc).

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js [hunk 5]
```diff
@@ -63,32 +67,64 @@ class SubscriptionDetails extends Component {
...
+              <Nav bsClass="nav nav-tabs">
+                <NavItem eventKey={1}>
+                  <div>{__('Details')}</div>
+                </NavItem>
+                <NavItem eventKey={2}>
+                  <div>{__('Enabled Products')}</div>
+                </NavItem>
+              </Nav>
+              <Grid bsClass="container-fluid">
+                <TabContent animation={false}>
```
Comment (amirfefer) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: using react router on tabs content may increase cohesion and reduce the snapshot size.

**Comment by akofink** [comment]
So... this is what it looks like for me:

image

I have no idea where the heck the 'Partition Table' column is coming from. Anyone?

@Rohoover I'm not convinced this data looks nice in a table. Most of the cells have quite long text. In addition, I think we might want to group by product name.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js (hunk 5 above)
Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: I'm wondering about this too since the tabs don't display new pages

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        formatters: [headerFormatter],
+      },
+      cell: {
+        formatters: [cellFormatter],
+      },
+      ...column_overrides,
+    }
+  };
+
+  const simple_column = (property, label) =>
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: These seem like they could be a good addition to the table related components.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  };
+
+  const enabledProductColumns = [
+    simple_column('name', 'Name'),
+    simple_column('product', 'Product'),
+    simple_column('content_url', 'Content Download URL'),
+    simple_column('gpg_url', 'GPG Key URL'),
+    simple_column('type', 'Repo Type'),
+    simple_column('enabled', 'Enabled'),
+  ];
```
Comment (johnpmitsch) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: We generally put the Table schema in a different file: i.e. https://github.com/Katello/katello/blob/master/webpack/scenes/Subscriptions/components/SubscriptionsTable/SubscriptionsTableSchema.js

However, since you created some nice helper functions, those would be a good thing to put those in our helper file `webpack/move_to_foreman/common/helpers.js`. If you do this, you could just import the `simple_column` function here and keep `enabledProductColumns` in this component since its pretty concise.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    column({
+      property: property,
+      header: {
+        label: __(label),
+        formatters: [headerFormatter],
+      },
+    });
+
+  const emptyStateData = () => ({
+    header: __('There are no Subscription Allocations to display'),
```
Comment (akofink) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: This is obviously just placeholder I copied from upstream subscriptions XD

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        return {
+          product: product.name,
+          enabled: content.enabled,
+          ...content.content,
+        };
+      })
+    );
+  };
+
+  const enabledProductColumns = [
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: We have been putting these in schema files to make the component itself less complicated.  See  [`UpstreamSubscriptionsTableSchema`](https://github.com/Katello/katello/blob/master/webpack/scenes/Subscriptions/UpstreamSubscriptions/UpstreamSubscriptionsTableSchema.js) for example.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js [hunk 10]
```diff
@@ -45,7 +58,7 @@ class SubscriptionDetails extends Component {
     }
 
     return (
-      <Grid bsClass="container-fluid">
+      <div>
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: Why this change?

**Comment by waldenraines** [comment]
> @Rohoover I'm not convinced this data looks nice in a table. Most of the cells have quite long text. In addition, I think we might want to group by product name.

What if we used [ListView](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-simple-expansion.html)?

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js (hunk 10 above)
Comment (akofink) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: This Grid moved to inside the TabContainer, on [line 91](https://github.com/Katello/katello/pull/7578/files#diff-d67a0043161dfe1a4d75f6f64d995cbfR91).

**Comment by akofink** [comment]
Updated to use ListView, still WIP for tests.

**Comment by waldenraines** [comment]
> Updated to use ListView, still WIP for tests. It looks like this:

@Rohoover thoughts on this?  To me it looks better than the table.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+                      <Row>Repo Type</Row>
+                      <Row>Enabled?</Row>
+                    </Col>
+                    <Col sm={9}>
+                      <Row>{content.content_url}</Row>
+                      <Row>{content.gpg_url}</Row>
+                      <Row>{content.type}</Row>
+                      <Row>{content.enabled ? __('yes') : __('no')}</Row>
+                    </Col>
+                  </Row>
```
Comment (johnpmitsch) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: You *could* break this out into a component, I'm a fan of having the Row part of a table/list be an individual component for readability. But I don't consider that a blocker, so I will leave that up to you :wink:

**Comment by johnpmitsch** [comment]
Looks good!

**Comment by Rohoover** [comment]
LGTM!

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js (hunk 11 above)
Comment (akofink) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: I agree

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+      product.available_content.map(c => (
+        {
+          enabled: c.enabled,
+          ...c.content,
+        }
+      ))
+    ),
+  }));
+
+  const ProductContent = ({ content }) => (
```
[the remark this answers is not in the document]
Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: I think it should live in its own file

[the remark this answers is not in the document]
Comment (akofink) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: Really? I'd only think that if we were going to use it elsewhere. It's specific to this one SubscriptionDetailEnabledProducts component. Is there some one component per file rule in React?

[the remark this answers is not in the document]
Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: React isn't too opinionated on this stuff. It doesn't matter once everything goes through webpack, its just for organization. I generally see one component per file (even if its not used elsewhere) unless its something that makes sense to group like `GreenButton`, `RedButton`, etc...

Its just my opinion that it should be in its own file but not a hard rule, so its really up to you, its not a problem to me if its stays here. :)

[the remark this answers is not in the document]
Comment (akofink) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailEnabledProducts.js]: I moved it to a separate file. :)

**Comment by akofink** [comment]
This should be good to go!

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js [hunk 13]
```diff
@@ -18,12 +19,22 @@ class SubscriptionDetails extends Component {
     // eslint-disable-next-line react/prop-types
     const routerParams = this.props.match.params;
     this.props.loadSubscriptionDetails(parseInt(routerParams.id, 10));
+    this.props.loadProducts({
+      subscription_id: parseInt(routerParams.id, 10),
```
Comment (johnpmitsch) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: I wonder if `parseInt(routerParams.id, 10)` can be moved to a constant somewhere since we keep repeating it in this file?

**Comment by johnpmitsch** [comment]
@akofink If you want to, but that may require a change in jenkins and then here in katello so unsure how long it will be (hopefully will be fixed soon). If you there is no difference in `npm run lint` on master vs your branch, you haven't introduced any new linting errors.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js (hunk 13 above)
Comment (waldenraines) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: That's a good thought, maybe a helper named `parseBaseTen()` or something?

**Comment by waldenraines** [comment]
> Let's wait for the eslint PR :)

I'd say go ahead and merge it.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js (hunk 13 above)
Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: I more meant for the subscription id that its getting, have a const like `subscriptionId` that can be used.

Comment (waldenraines) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: Ah, that makes more sense

**Comment by akofink** [comment]
There are some eslint failures in the tests that Jenkins isn't telling us about. Let me update.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js (hunk 13 above)
Comment (akofink) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: I'm willing to change this, but I also think that we might refactor this to come up with a solution that is reusable (as Walden was suggesting), especially since Foreman has to do these types of lookups as well.

Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: I would have to see what you mean - I don't see any value in abstracting out `parseInt` as that is a single function and `routerParams` is specific to this component  which is why I suggested refactoring just in this component

Comment (akofink) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: `routerParams` is just `this.props.match.params`. We're just parsing the passed in ID (which is a string that we need to make an integer) to the GET request. This attribute is common to all our APIs (afaik), so I'd think Foreman is doing something similar to solve this same problem.

Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: I'm open to suggestions! I was just thinking something like this (or similar) when I started the discussion:

```javascript
getSubscriptionId() {
  const routerParams = this.props.match.params;
  return parseInt(routerParams.id, 10);
}
```

**Comment by waldenraines** [comment]
> Updated for eslint

So now you can hit the big green button?

**Comment by akofink** [comment]
Yeah!
